lingpy/sequence/ngrams.py

```diff
@@ -0,0 +1,1441 @@
+    with Python 2.7.
+    """
+    # accumulate([1,2,3,4,5]) --> 1 3 6 10 15
+    # accumulate([1,2,3,4,5], operator.mul) --> 1 2 6 24 120
+    it = iter(iterable)
+    try:
+        total = next(it)
+    except StopIteration:
+        return
+    yield total
+    for element in it:
+        total = func(total, element)
+        yield total
+
+def _random_choices(population, weights=None, cum_weights=None, k=1):
+    """
+    Return a `k` sized list of elements chosen from `population` with
+    replacement and according to a list of weights.
+
+    If a `weights` sequence is specified, selections are made according to the
+    relative weights. Alternatively, if a `cum_weights` sequence is given, the
+    selections are made according to the cumulative weights. For example, the
+    relative weights `[10, 5, 30, 5]` are equivalent to the cumulative weights
+    `[10, 15, 45, 50]`. Internally, the relative weights are converted to the
+    cumulative weights before making selections, so supplying the cumulative
+    weights saves work.
+
+    This function is compatible with the random.choices() function available
+    in Python's standard library from version 3.6 on. It can be replaced by
+    the standard implementation once the version requirement is updated.
+
+    Parameters
+    ----------
+    population: list
+        A list of elements from which the element(s) will be drawn.
+
+    weights: list
+        A list of any numeric type with the relative weight of each element.
+        Either `weights` or `cum_weights` must be provided.
+
+    cum_weights: list
+        A list of any numeric type with the accumulated weight of each element.
+        Either `weights` or `cum_weights` must be provided.
+
+    k: int
+        The number of elements to be drawn, with replacement.
+
+    Returns
+    -------
+    sample: list
+        A list of elements randomly drawn according to the specified weights.
+    """
+
+    # Assert that (1) the population is not empty, (2) only one type of
+    # weight information is provided.
+    assert population, "Population must not be empty."
+    assert not all((weights, cum_weights)), \
+        "Either only weights or only cumulative weights must be provided."
+
+    # If cumulative weights were not provided, build them from `weights`.
+    if not cum_weights:
+        cum_weights = list(_accumulate(weights))
+
+    # Assert that the lengths of population and cumulative weights match.
+    assert len(population) == len(cum_weights), \
+        "Population and weight lengths do not match."
+
+    # Get a random number and see in which bin it falls. We need to use this
+    # logic which is a little more complex than something with randint()
+    # in order to allow for floating-point weights.
+    rnd = [random.uniform(0, cum_weights[-1]) for r in range(k)]
+    less_than = [[cw < r for cw in cum_weights] for r in rnd]
+
+    return [population[lt.index(False)] for lt in less_than]
+
```

Comment:
`_accumulate` and in particular `_random_choices` look like candidates for a utility module (and may be useful in other places, too).

Reply:
They are really only there due to py2, I am only using them also in py3 because I don't want to check for Python version each time (reminds of C's #ifdef...). I'll move them to `lingpy/utils.py` for the time being.